File: src/algorithms/recursion/permutations/solution1.go
package main

// Upper Bound: O(n^2*n!) time | O(n*n!) space
// Roughly: O(n*n!) time | O(n*n!) space
func GetPermutations(array []int) [][]int {
	permutations := [][]int{}
	permutationsHelper(array, []int{}, &permutations)
	return permutations
}

func permutationsHelper(array []int, currentPermutation []int, permutations *[][]int) {
	if len(array) == 0 && len(currentPermutation) != 0 {
		*permutations = append(*permutations, currentPermutation)
		return
	}
	for i := range array {
		newArray := make([]int, i)
		copy(newArray, array[:i])
		newArray = append(newArray, array[i+1:]...)
		newPermutation := make([]int, len(currentPermutation))
		copy(newPermutation, currentPermutation)
		newPermutation = append(newPermutation, array[i])
		permutationsHelper(newArray, newPermutation, permutations)
	}
}

File: src/algorithms/recursion/permutations/solution1.ts
// Upper Bound: O(n^2*n!) time | O(n*n!) space
// Roughly: O(n*n!) time | O(n*n!) space
export function getPermutations(array: number[]) {
  const permutations: number[][] = [];
  permutationsHelper(array, [], permutations);
  return permutations;
};

function permutationsHelper(
  array: number[],
  currentPermutation: number[],
  permutations: number[][]
) {
	if (!array.length && currentPermutation.length) {
		permutations.push(currentPermutation);
	} else {
		for (let i = 0; i < array.length; i++) {
			const newArray = array.slice(0, i).concat(array.slice(i + 1));
			const newPermutation = currentPermutation.concat([array[i]]);
			permutationsHelper(newArray, newPermutation, permutations)
		};
	};
};
